File: create_ekspertiz.py
from openpyxl import Workbook
from openpyxl.styles import (
    Font, PatternFill, Alignment, Border, Side, numbers
)
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.page import PageMargins

wb = Workbook()
ws = wb.active
ws.title = "Ön Hasar Ekspertiz"

# --- Stil tanımları ---
def thin_border(top=True, bottom=True, left=True, right=True):
    s = Side(style="thin", color="000000")
    n = Side(style=None)
    return Border(
        top=s if top else n,
        bottom=s if bottom else n,
        left=s if left else n,
        right=s if right else n,
    )

def thick_border():
    s = Side(style="medium", color="000000")
    return Border(top=s, bottom=s, left=s, right=s)

COLOR_HEADER_BG   = "1F3864"  # koyu lacivert
COLOR_HEADER_FG   = "FFFFFF"
COLOR_SECTION_BG  = "BDD7EE"  # açık mavi
COLOR_TABLE_HDR   = "2E75B6"  # mavi
COLOR_TABLE_HDR_FG= "FFFFFF"
COLOR_ROW_ALT     = "DEEAF1"  # zebra
COLOR_WHITE       = "FFFFFF"
COLOR_LABEL_BG    = "F2F2F2"

def style_cell(cell, bold=False, size=10, fg=None, bg=None,
               h_align="left", v_align="center", wrap=False, border=None):
    cell.font = Font(name="Calibri", bold=bold, size=size,
                     color=fg if fg else "000000")
    if bg:
        cell.fill = PatternFill("solid", fgColor=bg)
    cell.alignment = Alignment(horizontal=h_align, vertical=v_align,
                                wrap_text=wrap)
    if border:
        cell.border = border

def merge_style(ws, cell_range, value="", bold=False, size=10,
                fg=None, bg=None, h_align="left", v_align="center",
                wrap=False, border=None):
    ws.merge_cells(cell_range)
    first = ws[cell_range.split(":")[0]]
    first.value = value
    style_cell(first, bold=bold, size=size, fg=fg, bg=bg,
               h_align=h_align, v_align=v_align, wrap=wrap, border=border)
    # kenarlık tüm birleşik hücrelere
    if border:
        for row in ws[cell_range]:
            for c in row:
                c.border = border

# --- Sütun genişlikleri ---
# A  B  C  D  E  F  G  H  I
col_widths = [2, 22, 14, 14, 14, 12, 12, 12, 22]
for i, w in enumerate(col_widths, start=1):
    ws.column_dimensions[get_column_letter(i)].width = w

# --- Satır yükseklikleri (sonra tek tek ayarlanacak) ---
row = 1

# ===================== BAŞLIK =====================
ws.row_dimensions[row].height = 8
row += 1  # row 2

ws.row_dimensions[row].height = 36
merge_style(ws, f"B{row}:I{row}",
            value="ARAÇ ÖN HASAR EKSPERTİZ FORMU",
            bold=True, size=16, fg=COLOR_HEADER_FG, bg=COLOR_HEADER_BG,
            h_align="center", v_align="center",
            border=thin_border())
row += 1  # row 3

ws.row_dimensions[row].height = 18
# Rapor No / Tarih / Ekspertiz Yeri
for col, label, val_col in [("B", "Rapor No :", "C"), ("E", "Tarih :", "F"), ("G", "Ekspertiz Yeri :", "I")]:
    c = ws[f"{col}{row}"]
    c.value = label
    style_cell(c, bold=True, size=9, bg=COLOR_LABEL_BG, border=thin_border())
    vc = ws[f"{val_col}{row}"]
    style_cell(vc, size=9, border=thin_border())

ws.merge_cells(f"C{row}:D{row}")
ws.merge_cells(f"F{row}:F{row}")
ws.merge_cells(f"I{row}:I{row}")
row += 1  # row 4
ws.row_dimensions[row].height = 6
row += 1  # row 5

# ===================== ARAÇ BİLGİLERİ =====================
ws.row_dimensions[row].height = 18
merge_style(ws, f"B{row}:I{row}",
            value="  ARAÇ BİLGİLERİ",
            bold=True, size=10, fg="000000", bg=COLOR_SECTION_BG,
            h_align="left", v_align="center",
            border=thin_border())
row += 1

arac_fields = [
    [("Plaka", "C"), ("Marka / Model", "F")],
    [("Yıl", "C"), ("Renk", "F")],
    [("Şase No", "C"), ("Motor No", "F")],
    [("Kilometre", "C"), ("Yakıt Türü", "F")],
]

for pair in arac_fields:
    ws.row_dimensions[row].height = 18
    for label, val_end in pair:
        if val_end == "C":
            lbl_col, val_start, val_end2 = "B", "C", "D"
        else:
            lbl_col, val_start, val_end2 = "E", "F", "I"
        lc = ws[f"{lbl_col}{row}"]
        lc.value = label
        style_cell(lc, bold=True, size=9, bg=COLOR_LABEL_BG, border=thin_border())
        ws.merge_cells(f"{val_start}{row}:{val_end2}{row}")
        vc = ws[f"{val_start}{row}"]
        style_cell(vc, size=9, border=thin_border())
    row += 1

ws.row_dimensions[row].height = 6
row += 1

# ===================== SİGORTA BİLGİLERİ =====================
ws.row_dimensions[row].height = 18
merge_style(ws, f"B{row}:I{row}",
            value="  SİGORTA / POLİÇE BİLGİLERİ",
            bold=True, size=10, fg="000000", bg=COLOR_SECTION_BG,
            h_align="left", v_align="center",
            border=thin_border())
row += 1

sig_fields = [
    [("Sigorta Şirketi", "C"), ("Poliçe No", "F")],
    [("Hasar Tarihi", "C"), ("İhbar Tarihi", "F")],
    [("Dosya No", "C"), ("", "F")],
]

for pair in sig_fields:
    ws.row_dimensions[row].height = 18
    for label, val_end in pair:
        if val_end == "C":
            lbl_col, val_start, val_end2 = "B", "C", "D"
        else:
            lbl_col, val_start, val_end2 = "E", "F", "I"
        lc = ws[f"{lbl_col}{row}"]
        lc.value = label
        style_cell(lc, bold=True, size=9, bg=COLOR_LABEL_BG, border=thin_border())
        ws.merge_cells(f"{val_start}{row}:{val_end2}{row}")
        vc = ws[f"{val_start}{row}"]
        style_cell(vc, size=9, border=thin_border())
    row += 1

ws.row_dimensions[row].height = 6
row += 1

# ===================== HASAR TESPİT TABLOSU =====================
ws.row_dimensions[row].height = 18
merge_style(ws, f"B{row}:I{row}",
            value="  HASAR TESPİT TABLOSU (ÖN BÖLGE)",
            bold=True, size=10, fg="000000", bg=COLOR_SECTION_BG,
            h_align="left", v_align="center",
            border=thin_border())
row += 1

# Tablo başlıkları
tbl_headers = ["PARÇA", "HASAR YOK", "BOYALI", "DEĞİŞMİŞ", "HASARLI", "KOPUK/\nKIRIK", "NOTLAR"]
tbl_cols    = ["B", "C", "D", "E", "F", "G", "H"]  # H->I birleştirilecek
ws.row_dimensions[row].height = 28
for i, (col, hdr) in enumerate(zip(tbl_cols, tbl_headers)):
    c = ws[f"{col}{row}"]
    c.value = hdr
    style_cell(c, bold=True, size=9, fg=COLOR_TABLE_HDR_FG, bg=COLOR_TABLE_HDR,
               h_align="center", v_align="center", wrap=True, border=thin_border())
# Notlar sütunu B ile I arasında H:I
ws.merge_cells(f"H{row}:I{row}")
row += 1

parcalar = [
    "Ön Tampon",
    "Kaput",
    "Sağ Ön Çamurluk",
    "Sol Ön Çamurluk",
    "Ön Izgara",
    "Sağ Far",
    "Sol Far",
    "Sağ Ön Kapı",
    "Sol Ön Kapı",
    "Ön Cam",
    "Motor (Görsel)",
    "Radyatör (Görsel)",
    "Airbag",
    "Diğer",
]

for idx, parca in enumerate(parcalar):
    ws.row_dimensions[row].height = 16
    bg = COLOR_ROW_ALT if idx % 2 == 0 else COLOR_WHITE
    # Parça adı
    pc = ws[f"B{row}"]
    pc.value = parca
    style_cell(pc, size=9, bg=bg, border=thin_border())
    # Checkbox sütunlar (C-G) — boş, ortalı
    for col in ["C", "D", "E", "F", "G"]:
        c = ws[f"{col}{row}"]
        style_cell(c, size=9, bg=bg, h_align="center", border=thin_border())
    # Notlar (H:I)
    ws.merge_cells(f"H{row}:I{row}")
    nc = ws[f"H{row}"]
    style_cell(nc, size=9, bg=bg, border=thin_border())
    row += 1

ws.row_dimensions[row].height = 6
row += 1

# ===================== EKSPERTİZ NOTLARI =====================
ws.row_dimensions[row].height = 18
merge_style(ws, f"B{row}:I{row}",
            value="  EKSPERTİZ NOTLARI / GENEL GÖZLEMLER",
            bold=True, size=10, fg="000000", bg=COLOR_SECTION_BG,
            h_align="left", v_align="center",
            border=thin_border())
row += 1

ws.row_dimensions[row].height = 60
merge_style(ws, f"B{row}:I{row}",
            value="",
            bold=False, size=9, bg=COLOR_WHITE,
            h_align="left", v_align="top", wrap=True,
            border=thin_border())
row += 1

ws.row_dimensions[row].height = 6
row += 1

# ===================== İMZA BÖLÜMÜ =====================
ws.row_dimensions[row].height = 18
merge_style(ws, f"B{row}:I{row}",
            value="  İMZA",
            bold=True, size=10, fg="000000", bg=COLOR_SECTION_BG,
            h_align="left", v_align="center",
            border=thin_border())
row += 1

ws.row_dimensions[row].height = 50
# Araç Sahibi
ws.merge_cells(f"B{row}:E{row}")
ac = ws[f"B{row}"]
ac.value = "Araç Sahibi\nAd Soyad :\nİmza         :"
style_cell(ac, size=9, bg=COLOR_LABEL_BG, h_align="left", v_align="top",
           wrap=True, border=thin_border())
# Ekspertiz Uzmanı
ws.merge_cells(f"F{row}:I{row}")
ec = ws[f"F{row}"]
ec.value = "Ekspertiz Uzmanı\nAd Soyad :\nÜnvan        :\nİmza           :"
style_cell(ec, size=9, bg=COLOR_LABEL_BG, h_align="left", v_align="top",
           wrap=True, border=thin_border())
row += 1

ws.row_dimensions[row].height = 8

# ===================== YAZDIRILACAK ALAN =====================
last_data_row = row
ws.print_area = f"A1:{get_column_letter(9)}{last_data_row}"

# ===================== SAYFA AYARLARI =====================
ws.page_setup.paperSize   = ws.PAPERSIZE_A4
ws.page_setup.orientation = ws.ORIENTATION_PORTRAIT
ws.page_setup.fitToPage   = True
ws.page_setup.fitToWidth  = 1
ws.page_setup.fitToHeight = 1

ws.page_margins = PageMargins(
    left=0.4, right=0.4, top=0.5, bottom=0.5,
    header=0.2, footer=0.2
)

ws.sheet_properties.pageSetUpPr.fitToPage = True

# ===================== KAYDET =====================
out_path = r"C:\Users\idemi\Desktop\verdent-project\on_hasar_ekspertiz.xlsx"
wb.save(out_path)
print(f"Dosya kaydedildi: {out_path}")

File: create_ekspertiz.js
const ExcelJS = require('exceljs');
const path = require('path');

const wb = new ExcelJS.Workbook();
const ws = wb.addWorksheet('Ön Hasar Ekspertiz', {
  pageSetup: {
    paperSize: 9,
    orientation: 'portrait',
    fitToPage: true,
    fitToWidth: 1,
    fitToHeight: 1,
  }
});

// Sütunlar: A(margin) B(PNo-Sol) C(PAdı-Sol) D(Fiyat-Sol) E(sep) F(PNo-Sağ) G(PAdı-Sağ) H(Fiyat-Sağ) I(sep) J(InfoLabel) K(InfoValue) L(margin)
ws.columns = [
  { width: 1 },    // A
  { width: 5 },    // B - Parça No Sol
  { width: 14 },   // C - Parça Adı Sol
  { width: 8 },    // D - Fiyatı Sol
  { width: 1.5 },  // E - ayraç
  { width: 5 },    // F - Parça No Sağ
  { width: 14 },   // G - Parça Adı Sağ
  { width: 8 },    // H - Fiyatı Sağ
  { width: 1.5 },  // I - ayraç
  { width: 12 },   // J - Info Label / Hizmet Kalem
  { width: 9 },    // K - Info Value / Tutar
  { width: 1 },    // L - margin
];

const C = {
  darkBlue : 'FF1F3864',
  red      : 'FFCC0000',
  white    : 'FFFFFFFF',
  lightBlue: 'FFBDD7EE',
  tableHdr : 'FF2E75B6',
  altRow   : 'FFDEEAF1',
  labelBg  : 'FFF2F2F2',
  yellow   : 'FFFFFF99',
  black    : 'FF000000',
};

const thinSide = { style: 'thin', color: { argb: C.black } };
const thickSide = { style: 'medium', color: { argb: C.black } };

function borders(cell, thick = false) {
  const s = thick ? thickSide : thinSide;
  cell.border = { top: s, bottom: s, left: s, right: s };
}

function applyStyle(cell, { bold=false, size=9, fg=C.black, bg=null, h='left', v='middle', wrap=false, thick=false }={}) {
  cell.font = { name: 'Calibri', bold, size, color: { argb: fg } };
  if (bg) cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: bg } };
  cell.alignment = { horizontal: h, vertical: v, wrapText: wrap };
  borders(cell, thick);
}

// Hücre birleştir + değer + stil
function mc(r, c1, c2, val, opts={}) {
  if (c1 !== c2) ws.mergeCells(r, c1, r, c2);
  const cell = ws.getCell(r, c1);
  cell.value = val;
  applyStyle(cell, opts);
  return cell;
}

// Tek hücre
function sc(r, c, val, opts={}) {
  const cell = ws.getCell(r, c);
  cell.value = val;
  applyStyle(cell, opts);
  return cell;
}

let row = 1;

// ─────────────────────────────────────────────
// BAŞLIK SATIRI
// ─────────────────────────────────────────────
ws.getRow(row).height = 5; row++;

// Logo + Rapor No satırı
ws.getRow(row).height = 34;
mc(row, 1, 4, 'odak', { bold:true, size:20, fg:C.white, bg:C.darkBlue, h:'center', v:'middle', thick:true });
mc(row, 5, 8, '002138', { bold:true, size:20, fg:C.red, bg:'FFFAFAFA', h:'center', v:'middle', thick:true });
mc(row, 9, 12, 'www.odakotomariv.com\nTel: (0212) 449 13 13\nFax: (0212) 483 66 55', { size:7, fg:C.black, bg:'FFFAFAFA', h:'center', v:'middle', wrap:true });
row++;

// Adres satırı (sol)
ws.getRow(row).height = 13;
mc(row, 1, 8, 'Cevizlibağ Maltepe Mh. Londra Asfaltı Cad. Odak Plaza No:40/13A 34010 Zeytinburnu/İST.', { size:7.5, h:'center', v:'middle' });
mc(row, 9, 12, '', { bg:C.labelBg });
row++;

// Uyarı satırı (sol) + SN (sağ)
ws.getRow(row).height = 12;
mc(row, 1, 8, 'BU EKSPERTİZ BİR ÖN TAHMİNDİR, KESİN SONUÇ ONARIM NETİCESİNDE BELLİ OLUR. ARKASINDAKI ONARIM GENEL ŞARTLARINA DİKKAT EDİNİZ.',
  { bold:true, size:7, h:'center', v:'middle' });
sc(row, 9, 'SN.', { bold:true, size:8, bg:C.labelBg });
mc(row, 10, 12, '', { bg:C.labelBg });
row++;

// ─────────────────────────────────────────────
// CHECKBOX (sol) + ARAÇ BİLGİLERİ (sağ) bölümü
// ─────────────────────────────────────────────
ws.getRow(row).height = 16;
// Checkbox başlıkları
mc(row, 1, 2, 'EKİPMAN', { bold:true, size:8, bg:C.lightBlue, h:'center' });
sc(row, 3, 'VAR', { bold:true, size:8, bg:C.lightBlue, h:'center' });
sc(row, 4, 'YOK', { bold:true, size:8, bg:C.lightBlue, h:'center' });
mc(row, 5, 8, '', { bg:C.lightBlue });
// Araç bilgileri başlık
sc(row, 9, 'TELEFON', { bold:true, size:8, bg:C.labelBg });
mc(row, 10, 12, '', { size:8 });
row++;

const checkboxRows = [
  'SİGORTA', 'AIRBAG', 'KLİMA', 'SUNROOF',
  'R.E.T.', 'İMMOBİLİZER', 'ALARM', 'DERİ KOLTUK',
];
const vehicleFields = [
  'PLAKA NO', 'ARAÇ GELİŞ TARİHİ :', 'ŞASİ NO :',
  'MOTOR NO :', 'ARAÇ MODELİ :', 'KM/MİL :',
  'ARAÇ RENGİ :', 'TESBİT EDEN :',
];

for (let i = 0; i < 8; i++) {
  ws.getRow(row).height = 15;
  const bg = i % 2 === 0 ? C.altRow : C.white;
  mc(row, 1, 2, checkboxRows[i], { size:8, bg });
  sc(row, 3, '☐', { size:10, h:'center', bg });
  sc(row, 4, '☐', { size:10, h:'center', bg });
  mc(row, 5, 8, '', { bg });
  sc(row, 9, vehicleFields[i], { bold:true, size:8, bg:C.labelBg });
  mc(row, 10, 12, '', { size:8 });
  row++;
}

// KM satırı (son checkbox)
ws.getRow(row).height = 15;
mc(row, 1, 2, 'KM.', { size:8, bg:C.altRow });
mc(row, 3, 4, '', { size:8, bg:C.altRow, h:'center' });
mc(row, 5, 8, '', { bg:C.altRow });
mc(row, 9, 12, '', { bg:C.lightBlue });
row++;

ws.getRow(row).height = 5; row++;

// ─────────────────────────────────────────────
// PARÇA TABLOSU
// ─────────────────────────────────────────────
// TOPLAM gösterim satırı (üst)
ws.getRow(row).height = 18;
mc(row, 1, 3, 'TOPLAM', { bold:true, size:9, fg:C.white, bg:C.darkBlue, h:'right', v:'middle' });
sc(row, 4, '', { bold:true, size:9, bg:C.yellow, h:'right' });
mc(row, 5, 7, 'TOPLAM', { bold:true, size:9, fg:C.white, bg:C.darkBlue, h:'right', v:'middle' });
sc(row, 8, '', { bold:true, size:9, bg:C.yellow, h:'right' });
mc(row, 9, 12, '', { bg:C.darkBlue });
row++;

// Tablo sütun başlıkları
ws.getRow(row).height = 18;
sc(row, 1, 'PARÇA NO', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
sc(row, 2, 'PARÇA ADI', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
sc(row, 3, '', { bg:C.tableHdr });
sc(row, 4, 'FİYATI', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
sc(row, 5, 'PARÇA NO', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
sc(row, 6, 'PARÇA ADI', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
sc(row, 7, '', { bg:C.tableHdr });
sc(row, 8, 'FİYATI', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
mc(row, 9, 10, 'HİZMET KALEMİ', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
mc(row, 11, 12, 'TUTAR', { bold:true, size:8, fg:C.white, bg:C.tableHdr, h:'center' });
row++;

// Hizmet kalemleri (sağ sütunda gösterilecek)
const serviceItems = [
  'KAPORTA', 'BOYA', 'TİRİM-DÖŞEME', 'KİLİT-CAM',
  'MEKANİK', 'ELEKTRİK', 'CELET', 'KLİMA GAZİ',
  'ROTBALANS', 'TOP İŞÇİL',
];

// 20 parça satırı
for (let i = 0; i < 20; i++) {
  ws.getRow(row).height = 15;
  const bg = i % 2 === 0 ? C.altRow : C.white;

  sc(row, 1, '', { bg, h:'center' });
  sc(row, 2, '', { bg });
  sc(row, 3, '', { bg });
  sc(row, 4, '', { bg, h:'right' });
  sc(row, 5, '', { bg, h:'center' });
  sc(row, 6, '', { bg });
  sc(row, 7, '', { bg });
  sc(row, 8, '', { bg, h:'right' });

  if (i < serviceItems.length) {
    mc(row, 9, 10, serviceItems[i], { bold:true, size:8, bg:C.labelBg });
    mc(row, 11, 12, '', { size:8, h:'right' });
  } else if (i === serviceItems.length) {
    mc(row, 9, 10, 'YED. PARÇA TUTARI', { bold:true, size:8, bg:C.lightBlue });
    mc(row, 11, 12, '', { size:8, h:'right', bg:C.lightBlue });
  } else if (i === serviceItems.length + 1) {
    mc(row, 9, 10, 'TAHMİNİ TUTARI', { bold:true, size:8, fg:C.white, bg:C.darkBlue });
    mc(row, 11, 12, '', { size:9, h:'right', bg:C.yellow });
  } else {
    mc(row, 9, 12, '', { bg: i % 2 === 0 ? C.altRow : C.white });
  }
  row++;
}

// TOPLAM alt satırı
ws.getRow(row).height = 18;
mc(row, 1, 3, 'TOPLAM', { bold:true, size:9, fg:C.white, bg:C.darkBlue, h:'right', v:'middle' });
sc(row, 4, '', { bold:true, size:9, bg:C.yellow, h:'right', thick:true });
mc(row, 5, 7, 'TOPLAM', { bold:true, size:9, fg:C.white, bg:C.darkBlue, h:'right', v:'middle' });
sc(row, 8, '', { bold:true, size:9, bg:C.yellow, h:'right', thick:true });
mc(row, 9, 10, 'Müşteri:', { bold:true, size:9, bg:C.labelBg });
mc(row, 11, 12, '', { size:9 });
row++;

ws.getRow(row).height = 5; row++;

// ─────────────────────────────────────────────
// İMZA / EKSPER SATIRI
// ─────────────────────────────────────────────
ws.getRow(row).height = 45;
mc(row, 1, 8, 'Adı - Soyadı / İmza', { size:8, v:'bottom', h:'center' });
mc(row, 9, 10, 'Eksper Sn.:\nTelefon :\nZiyaret Tarihi :\nSon.Ziy.Tarihi :', { size:8, bg:C.labelBg, v:'top', wrap:true });
mc(row, 11, 12, '', { v:'top' });
row++;

ws.getRow(row).height = 5;

// ─────────────────────────────────────────────
// YAZDIRMA AYARLARI
// ─────────────────────────────────────────────
ws.pageSetup.printArea = `A1:L${row}`;
ws.pageSetup.margins = { left:0.4, right:0.4, top:0.5, bottom:0.5, header:0.2, footer:0.2 };
ws.sheetProperties = { pageSetUpPr: { fitToPage: true } };

const outPath = path.join(__dirname, 'on_hasar_ekspertiz.xlsx');
wb.xlsx.writeFile(outPath).then(() => {
  console.log('Kaydedildi:', outPath);
}).catch(err => console.error(err));
